keywords/CouchbaseServer.py

```diff
@@ -250,3 +267,212 @@ def get_server_docs_with_prefix(self, url, bucket, prefix):
                 found_ids.append(row["$1"]["id"])
 
         return found_ids
+
+    def _wait_for_rebalance_complete(self, url):
+        """
+        Polls couchbase server tasks endpoint for any running rebalances.
+        Exits when no rebalances are in running state
+
+        /pools/default/tasks format:
+        [
+            {
+                "type": "rebalance",
+                "status": "running",
+                ...
+            }
+        ]
+        """
+
+        start = time.time()
+        while True:
+
+            if time.time() - start > keywords.constants.REBALANCE_TIMEOUT:
+                raise Exception("wait_for_rebalance_complete: TIMEOUT")
+
+            resp = self._session.get("{}/pools/default/tasks".format(url))
+            log_r(resp)
+            resp.raise_for_status()
+
+            resp_obj = resp.json()
+
+            done_rebalacing = True
+            for task in resp_obj:
+                # loop through each task and see if any rebalance tasks are running
+                task_type = task["type"]
+                task_status = task["status"]
+                log_info("{} is {}".format(task_type, task_status))
+                if task_type == "rebalance" and task_status == "running":
+                    done_rebalacing = False
+
+            if done_rebalacing:
+                break
+
+            time.sleep(1)
+
+    def add_node(self, server_to_add):
+        """
+        Add the server_to_add to a Couchbase Server cluster
+        """
+
+        if not isinstance(server_to_add, CouchbaseServer):
+            raise TypeError("'server_to_add' must be a 'CouchbaseServer'")
+
+        log_info("Adding server node {} to cluster ...".format(server_to_add))
+        data = "hostname={}&user=Administrator&password=password&services=kv".format(
+            server_to_add.host
+        )
+
+        # HACK: Retries added to handle the following scenario.
+        # 1. A node has been rebalanced out of a cluster
+        # 2. Try to rebalance the node back in immediately.
+        # The node may not be in a ready state. There in no property exposed on the
+        #   REST API to poll so a retry will have to do
+        start = time.time()
```

Comment:
I'm not really clear on the scenario described above .. can you spell this out a little more?  Also describe how the hack fixes it.